File: macos/Sources/RelicCore/RelicData.swift
import Foundation

public struct RelicInfo: Codable, Hashable, Sendable, Identifiable {
    public let id: Int
    public let name: String
    public let color: Int
    public let deep: Bool
    public let slots: [Int]
    public let curseSlots: [Int]

    public init(id: Int, name: String, color: Int, deep: Bool, slots: [Int], curseSlots: [Int]) {
        self.id = id
        self.name = name
        self.color = color
        self.deep = deep
        self.slots = slots
        self.curseSlots = curseSlots
    }
}

public struct ExtraAffix: Codable, Hashable, Sendable {
    public let effectID: Int
    public let name: String
    public let sortID: Int
    public let compatibilityID: Int

    private enum CodingKeys: String, CodingKey {
        case effectID = "effectId"
        case name
        case sortID = "sortId"
        case compatibilityID = "compatibilityId"
    }

    public init(effectID: Int, name: String, sortID: Int, compatibilityID: Int) {
        self.effectID = effectID
        self.name = name
        self.sortID = sortID
        self.compatibilityID = compatibilityID
    }
}

public struct RelicCatalog: Codable, Sendable {
    public let relicsSchemaVersion: Int
    public let gameVersion: String
    public let dataVersion: String
    public let generatedAt: String
    public let sources: [CatalogSource]
    public let relics: [RelicInfo]
    public let pools: [String: [Int]]
    public let extraAffixes: [ExtraAffix]

    public init(
        relicsSchemaVersion: Int = 1,
        gameVersion: String,
        dataVersion: String,
        generatedAt: String,
        sources: [CatalogSource],
        relics: [RelicInfo],
        pools: [String: [Int]],
        extraAffixes: [ExtraAffix]
    ) {
        self.relicsSchemaVersion = relicsSchemaVersion
        self.gameVersion = gameVersion
        self.dataVersion = dataVersion
        self.generatedAt = generatedAt
        self.sources = sources
        self.relics = relics
        self.pools = pools
        self.extraAffixes = extraAffixes
    }
}

public enum RelicDataError: LocalizedError {
    case unreadable
    case unsupportedSchema(Int)

    public var errorDescription: String? {
        switch self {
        case .unreadable: return "无法读取遗物数据文件"
        case .unsupportedSchema(let version): return "不支持的遗物数据版本：\(version)"
        }
    }
}

public enum RelicDataLoader {
    public static func load(from url: URL) throws -> RelicCatalog {
        guard let data = try? Data(contentsOf: url) else {
            throw RelicDataError.unreadable
        }
        return try load(from: data)
    }

    public static func load(from data: Data) throws -> RelicCatalog {
        guard let catalog = try? JSONDecoder().decode(RelicCatalog.self, from: data) else {
            throw RelicDataError.unreadable
        }
        try validate(catalog)
        return catalog
    }

    public static func validate(_ catalog: RelicCatalog) throws {
        guard catalog.relicsSchemaVersion == 1 else {
            throw RelicDataError.unsupportedSchema(catalog.relicsSchemaVersion)
        }
    }
}

File: android/gamedata/src/main/kotlin/com/nightreign/relicchecker/gamedata/save/RelicData.kt
package com.nightreign.relicchecker.gamedata.save

import com.nightreign.relicchecker.catalog.AffixCatalog
import com.nightreign.relicchecker.gamedata.GameDataJson
import com.nightreign.relicchecker.gamedata.GameDataKey
import kotlinx.serialization.Serializable

// 遗物物品表（nightreign-relics-v1.03.4.json）的存档检查专用解析。
//
// 词条反查车道也会解析同一个文件，但两边各写一份 DTO（并行车道不能互相依赖未合并的代码），
// 合并时再统一。这里只声明审计要用的字段：sources 之类的说明字段不声明，解码时跳过。

@Serializable
internal data class RelicsFileDto(
    val relicsSchemaVersion: Int? = null,
    val gameVersion: String = "",
    val dataVersion: String = "",
    val relics: List<RelicDto> = emptyList(),
    val pools: Map<String, List<Long>> = emptyMap(),
    val extraAffixes: List<ExtraAffixDto> = emptyList(),
)

@Serializable
internal data class RelicDto(
    val id: Int? = null,
    val name: String = "",
    val color: Int = -1,
    val deep: Boolean = false,
    val slots: List<Int> = emptyList(),
    val curseSlots: List<Int> = emptyList(),
)

@Serializable
internal data class ExtraAffixDto(
    val effectId: Long? = null,
    val name: String = "",
    val sortId: Int = 0,
    val compatibilityId: Int = -1,
)

/** 遗物表里的一件遗物。[slots] / [curseSlots] 恒为 3 位，缺失补 -1（与 Windows 端的防御一致）。 */
data class RelicInfo(
    val id: Int,
    val name: String,
    val color: Int,
    val deep: Boolean,
    val slots: List<Int>,
    val curseSlots: List<Int>,
)

/** 词条库之外的补充词条（AttachEffectParam 全量补充）。 */
data class ExtraAffix(
    val effectId: Long,
    val name: String,
    val sortId: Int,
    val compatibilityId: Int,
)

/** 解析好的遗物表：遗物、槽池（effectId 集合）与补充词条。 */
class RelicDataSet(
    val gameVersion: String,
    val dataVersion: String,
    val relics: List<RelicInfo>,
    val pools: Map<Int, Set<Long>>,
    val extraAffixes: List<ExtraAffix>,
) {
    /** 同一 ID 出现多次时取第一条（与桌面端 uniquingKeysWith first 一致）。 */
    val relicsById: Map<Int, RelicInfo> = LinkedHashMap<Int, RelicInfo>().also { map ->
        relics.forEach { map.putIfAbsent(it.id, it) }
    }

    companion object {
        fun parse(text: String): RelicDataSet {
            val dto = GameDataJson.decode<RelicsFileDto>(GameDataKey.RELICS, text)
            GameDataJson.requireVersion(GameDataKey.RELICS, dto.relicsSchemaVersion)
            return RelicDataSet(
                gameVersion = dto.gameVersion,
                dataVersion = dto.dataVersion,
                relics = dto.relics.mapNotNull { relic ->
                    val id = relic.id ?: return@mapNotNull null
                    RelicInfo(
                        id = id,
                        name = relic.name,
                        color = relic.color,
                        deep = relic.deep,
                        slots = relic.slots.triple(),
                        curseSlots = relic.curseSlots.triple(),
                    )
                },
                pools = dto.pools.entries.mapNotNull { (key, members) ->
                    key.toIntOrNull()?.let { it to members.toHashSet() as Set<Long> }
                }.toMap(),
                extraAffixes = dto.extraAffixes.mapNotNull { extra ->
                    val effectId = extra.effectId ?: return@mapNotNull null
                    ExtraAffix(effectId, extra.name, extra.sortId, extra.compatibilityId)
                },
            )
        }

        private fun List<Int>.triple(): List<Int> = List(3) { getOrNull(it) ?: -1 }
    }
}

/**
 * 存档检查页用的全部静态数据：遗物表 + 审计上下文 + 词条说明。
 *
 * 在 `rememberGameData` 的后台线程里一次建好（索引约两千词条、六百个槽池），
 * 页面与「对比另一份存档」共用同一份，不再每份存档重复建索引。
 */
class SaveAuditData(
    val relicData: RelicDataSet,
    val context: RelicAuditContext,
    /** effectId → 词条说明（只收录词条库里非空的 explanation）。 */
    val explanations: Map<Long, String>,
) {
    companion object {
        /** rememberGameData 的解析器标识：返回类型为 [SaveAuditData]。 */
        const val PARSER_ID = "save.audit.v1"

        fun build(text: String, catalog: AffixCatalog): SaveAuditData = build(RelicDataSet.parse(text), catalog)

        fun build(relicData: RelicDataSet, catalog: AffixCatalog): SaveAuditData = SaveAuditData(
            relicData = relicData,
            context = RelicAuditContext(catalog, relicData),
            explanations = SaveAuditPipeline.explanations(catalog),
        )
    }
}
